### PR TITLE
Faster in-memory weight transfer for transformers-neuronx

### PR DESCRIPTION
**Pull request by Lokiiiiii**
## Overview ##

This PR improves the latency of transferring model weights from the HF representation to the INF2 representation by doing it in-memory as opposed to using the disk as a buffer.  For OPT-66B on inf2.48xl, this improves the latency of weight transfer from 840s to 5s. 

## Description ## 

Splitting the model into small pieces on the disk and then loading it, reduces CPU memory requirements. However, this operation demands high IOPS from the disk. Typical EBS volumes used by EC2 (gp3) and SageMaker (gp2) do not provide high IOPS by default. This leads to EBS throttling, which manifests as a massive latency to transfer model weights (upto 1400 secs).

However, Large Language Models are typically used with large instances with large CPU memory. As a result, for instances like inf2.24xl, inf2.48xl, trn1.32xl, loading from disk is wasteful in terms of instance hours. Direct in-memory transfer is much faster (<5s) in these cases.

This PR modifies the load_model code path to first try in-memory loading and then fall back to disk loading in the event of an overflow.

## Testing ##
Tested in EC2 on inf2.48xl and inf2.24xl with OPT-66B and OPT-13B

### CODE REVIEW

File: engines/python/setup/djl_python/transformers_neuronx.py [hunk 1]
```diff
@@ -122,10 +119,39 @@ def load_inf2_model(self, model_type, load_path):
             n_positions=self.n_positions,
             unroll=self.unroll)
 
+    def load_inf2_model_from_memory(self, model_type):
```
Comment (lanking520) [inline engines/python/setup/djl_python/transformers_neuronx.py:122]: what is making a difference with above one? Just an extra step to load with low memory?

Comment (lanking520) [reply engines/python/setup/djl_python/transformers_neuronx.py:122]: does this work as the same as we are passing in `low_cpu_mem_usage=True`

Comment (Lokiiiiii) [reply engines/python/setup/djl_python/transformers_neuronx.py:122]: It provides an alternative to the model split to disk and load from disk workflow.
It directly transfers the model weights in-memory.

Comment (Lokiiiiii) [reply engines/python/setup/djl_python/transformers_neuronx.py:122]: More details provided in the CR description

File: engines/python/setup/djl_python/transformers_neuronx.py [hunk 2]
```diff
@@ -122,10 +119,39 @@ def load_inf2_model(self, model_type, load_path):
...
+        else:
+            model = MODEL_TYPE_TO_MODEL[model_type](
+                self.model.config,
+                batch_size=self.batch_size,
+                amp=self.amp,
+                tp_degree=self.tensor_parallel_degree,
+                n_positions=self.n_positions,
+                unroll=self.unroll)
+        model.load_state_dict_low_memory(self.model.state_dict())
+        return model.load_state_dict_low_memory(self.model.state_dict())
```
[the remark this answers is not in the document]
Comment (Lokiiiiii) [reply engines/python/setup/djl_python/transformers_neuronx.py]: Typo ! Fixed in new revision.

**Comment by Lokiiiiii** [comment]
Additionally tested `facebook/opt-1.3b` and `OpenAssistant/llama2-13b-orca-8k-3319` on inf2.24xl (EC2)

**Review by lanking520 — APPROVED** [review]

**Pull request merged**